File: typescript/src/bibtex.ts
import { Reference, ReferenceInit } from "./reference";

export class BibTeXError extends Error {}

export function parseBibtexEntry(entry: string): Reference {
    const text = entry.trim();
    if (!text) {
        throw new BibTeXError("Empty BibTeX entry");
    }
    if (!text.startsWith("@")) {
        throw new BibTeXError("BibTeX entry must start with '@'");
    }

    const typeEnd = text.indexOf("{");
    if (typeEnd === -1) {
        throw new BibTeXError("Missing opening brace for entry");
    }
    const entryType = text.slice(1, typeEnd).trim();
    if (!entryType) {
        throw new BibTeXError("Entry type is missing");
    }

    let remainder = text.slice(typeEnd + 1).trim();
    if (!remainder.endsWith("}")) {
        throw new BibTeXError("BibTeX entry must end with '}'");
    }
    remainder = remainder.slice(0, -1).trim();
    const firstComma = remainder.indexOf(",");
    if (firstComma === -1) {
        throw new BibTeXError("BibTeX entry is missing fields");
    }

    const citeKey = remainder.slice(0, firstComma).trim();
    if (!citeKey) {
        throw new BibTeXError("Entry cite key is missing");
    }
    const fieldBlob = remainder.slice(firstComma + 1);

    const fields = parseFields(fieldBlob);
    return referenceFromFields(entryType, citeKey, fields);
}

export function referenceToBibtex(reference: Reference): string {
    if (!reference.entryType.trim()) {
        throw new BibTeXError("Reference entry_type is required for BibTeX output");
    }
    if (!reference.citeKey.trim()) {
        throw new BibTeXError("Reference cite_key is required for BibTeX output");
    }

    const fields = reference.mergedFields();
    const order = [
        "author",
        "title",
        "journal",
        "booktitle",
        "publisher",
        "year",
        "volume",
        "number",
        "pages",
        "doi",
        "url",
    ];

    const orderedItems: Array<[string, string]> = [];
    for (const key of order) {
        const value = fields[key];
        if (value) {
            orderedItems.push([key, value]);
            delete fields[key];
        }
    }
    for (const key of Object.keys(fields).sort()) {
        orderedItems.push([key, fields[key]]);
    }

    if (orderedItems.length === 0) {
        return `@${reference.entryType}{${reference.citeKey}}`;
    }

    const body = orderedItems.map(([key, value]) => `  ${key} = {${value}}`).join(",\n");
    return `@${reference.entryType}{${reference.citeKey},\n${body}\n}`;
}

function parseFields(blob: string): Record<string, string> {
    const fields: Record<string, string> = {};
    let idx = 0;
    const length = blob.length;

    while (idx < length) {
        idx = consumeWhitespace(blob, idx);
        if (idx >= length) break;

        const [name, afterName] = consumeFieldName(blob, idx);
        idx = consumeWhitespace(blob, afterName);
        idx = expectChar(blob, idx, "=", `Field '${name}' misses '=' sign`);
        idx = consumeWhitespace(blob, idx);

        const [value, afterValue] = consumeValue(blob, idx);
        fields[name] = cleanValue(value);
        idx = consumeDelimiter(blob, afterValue);
    }

    return fields;
}

function consumeValue(text: string, start: number): [string, number] {
    if (start >= text.length) {
        return ["", start];
    }
    const char = text[start];
    if (char === "{") {
        return consumeBracedValue(text, start);
    }
    if (char === '"') {
        return consumeQuotedValue(text, start);
    }
    return consumeBareValue(text, start);
}

function consumeBracedValue(text: string, start: number): [string, number] {
    let depth = 0;
    let idx = start + 1;
    while (idx < text.length) {
        const current = text[idx];
        if (current === "{") {
            depth += 1;
        } else if (current === "}") {
            if (depth === 0) {
                return [text.slice(start + 1, idx), idx + 1];
            }
            depth -= 1;
        }
        idx += 1;
    }
    throw new BibTeXError("Missing closing brace in value");
}

function consumeQuotedValue(text: string, start: number): [string, number] {
    let idx = start + 1;
    let escaped = false;
    while (idx < text.length) {
        const current = text[idx];
        if (current === '"' && !escaped) {
            return [text.slice(start + 1, idx), idx + 1];
        }
        escaped = current === "\\" && !escaped;
        if (current !== "\\") {
            escaped = false;
        }
        idx += 1;
    }
    throw new BibTeXError("Missing closing quote in value");
}

function consumeBareValue(text: string, start: number): [string, number] {
    let idx = start;
    while (idx < text.length && !",\n\r".includes(text[idx])) {
        idx += 1;
    }
    return [text.slice(start, idx).trim(), idx];
}

function cleanValue(value: string): string {
    return value.replace(/\s+/g, " ").trim();
}

function consumeWhitespace(text: string, start: number): number {
    let idx = start;
    while (idx < text.length && (/\s/.test(text[idx]) || text[idx] === ",")) {
        idx += 1;
    }
    return idx;
}

function consumeFieldName(text: string, start: number): [string, number] {
    let idx = start;
    while (idx < text.length && /[A-Za-z0-9_-]/.test(text[idx])) {
        idx += 1;
    }
    const name = text.slice(start, idx).toLowerCase();
    if (!name) {
        throw new BibTeXError("Field name missing in BibTeX entry");
    }
    return [name, idx];
}

function expectChar(text: string, idx: number, expected: string, error: string): number {
    if (idx >= text.length || text[idx] !== expected) {
        throw new BibTeXError(error);
    }
    return idx + 1;
}

function consumeDelimiter(text: string, idx: number): number {
    let cursor = idx;
    while (cursor < text.length && /\s/.test(text[cursor])) {
        cursor += 1;
    }
    if (cursor < text.length && text[cursor] === ",") {
        cursor += 1;
    }
    return cursor;
}

function referenceFromFields(entryType: string, citeKey: string, fields: Record<string, string>): Reference {
    const authorsField = fields["author"] ?? "";
    const authors = splitAuthors(authorsField);
    const standardKeys = new Set([
        "author",
        "title",
        "journal",
        "booktitle",
        "publisher",
        "year",
        "volume",
        "number",
        "pages",
        "doi",
        "url",
    ]);
    const extra: Record<string, string> = {};
    for (const [key, value] of Object.entries(fields)) {
        if (!standardKeys.has(key)) {
            extra[key] = value;
        }
    }
    const init: ReferenceInit = {
        entryType,
        citeKey,
        title: fields["title"],
        authors,
        journal: fields["journal"],
        booktitle: fields["booktitle"],
        publisher: fields["publisher"],
        year: fields["year"],
        volume: fields["volume"],
        issue: fields["number"],
        pages: fields["pages"],
        doi: fields["doi"],
        url: fields["url"],
        extraFields: extra,
    };
    return new Reference(init);
}

function splitAuthors(raw: string): string[] {
    if (!raw) return [];
    return raw.split(" and ").map((author) => author.trim()).filter(Boolean);
}

File: python/src/transtex/bibtex.py
"""Utility helpers for parsing and writing BibTeX entries."""
from __future__ import annotations

from typing import Dict, Tuple

from .reference import Reference


class BibTeXError(ValueError):
    """Raised when the library cannot parse a BibTeX entry."""


def parse_bibtex_entry(entry: str) -> Reference:
    """Parse a single BibTeX entry into a :class:`Reference`."""
    text = entry.strip()
    if not text:
        raise BibTeXError("Empty BibTeX entry")
    if not text.startswith("@"):
        raise BibTeXError("BibTeX entry must start with '@'")

    type_end = text.find("{")
    if type_end == -1:
        raise BibTeXError("Missing opening brace for entry")
    entry_type = text[1:type_end].strip()
    if not entry_type:
        raise BibTeXError("Entry type is missing")

    remainder = text[type_end + 1 :].strip()
    if not remainder.endswith("}"):
        raise BibTeXError("BibTeX entry must end with '}'")
    remainder = remainder[:-1].strip()
    if "," not in remainder:
        raise BibTeXError("BibTeX entry is missing fields")

    cite_key, field_blob = remainder.split(",", 1)
    cite_key = cite_key.strip()
    if not cite_key:
        raise BibTeXError("Entry cite key is missing")

    fields = _parse_fields(field_blob)
    return _reference_from_fields(entry_type, cite_key, fields)


def reference_to_bibtex(reference: Reference) -> str:
    """Serialize a :class:`Reference` back into a BibTeX entry."""
    fields = reference.merged_fields()
    order = [
        "author",
        "title",
        "journal",
        "booktitle",
        "publisher",
        "year",
        "volume",
        "number",
        "pages",
        "doi",
        "url",
    ]
    ordered_items = []
    for key in order:
        value = fields.pop(key, None)
        if value:
            ordered_items.append((key, value))
    for key in sorted(fields.keys()):
        ordered_items.append((key, fields[key]))

    if not ordered_items:
        body = ""
    else:
        field_lines = [f"  {key} = {{{value}}}" for key, value in ordered_items]
        body = ",\n".join(field_lines)
        body = "\n" + body + "\n"

    return f"@{reference.entry_type}{{{reference.cite_key},{body}}}"


def _parse_fields(blob: str) -> Dict[str, str]:
    fields: Dict[str, str] = {}
    idx = 0
    length = len(blob)
    while idx < length:
        while idx < length and blob[idx] in "\n\r\t ,":
            idx += 1
        if idx >= length:
            break

        start = idx
        while idx < length and (blob[idx].isalnum() or blob[idx] in "_-"):
            idx += 1
        name = blob[start:idx].lower()
        if not name:
            raise BibTeXError("Field name missing in BibTeX entry")

        while idx < length and blob[idx].isspace():
            idx += 1
        if idx >= length or blob[idx] != "=":
            raise BibTeXError(f"Field '{name}' misses '=' sign")
        idx += 1
        while idx < length and blob[idx].isspace():
            idx += 1
        value, idx = _consume_value(blob, idx)
        cleaned = _clean_value(value)
        fields[name] = cleaned

        while idx < length and blob[idx].isspace():
            idx += 1
        if idx < length and blob[idx] == ",":
            idx += 1
    return fields


def _consume_value(text: str, start: int) -> Tuple[str, int]:
    if start >= len(text):
        return "", start

    char = text[start]
    if char == "{":
        depth = 0
        idx = start + 1
        while idx < len(text):
            current = text[idx]
            if current == "{" :
                depth += 1
            elif current == "}":
                if depth == 0:
                    return text[start + 1 : idx], idx + 1
                depth -= 1
            idx += 1
        raise BibTeXError("Missing closing brace in value")

    if char == '"':
        idx = start + 1
        escaped = False
        while idx < len(text):
            current = text[idx]
            if current == '"' and not escaped:
                return text[start + 1 : idx], idx + 1
            escaped = current == "\\" and not escaped
            if current != "\\":
                escaped = False
            idx += 1
        raise BibTeXError("Missing closing quote in value")

    idx = start
    while idx < len(text) and text[idx] not in ",\n\r":
        idx += 1
    return text[start:idx].strip(), idx


def _clean_value(value: str) -> str:
    collapsed = " ".join(value.replace("\n", " ").split())
    return collapsed.strip()


def _reference_from_fields(entry_type: str, cite_key: str, fields: Dict[str, str]) -> Reference:
    authors_field = fields.get("author", "")
    authors = _split_authors(authors_field)
    normalized_entry = entry_type.lower()
    number_value = fields.get("number")
    issue = None
    report_number = None
    if number_value:
        if fields.get("journal") or normalized_entry in {"article", "inproceedings", "incollection"}:
            issue = number_value
        elif normalized_entry in {"techreport", "report"} or fields.get("institution"):
            report_number = number_value
        else:
            issue = number_value
    standard_keys = {
        "author",
        "title",
        "journal",
        "booktitle",
        "publisher",
        "address",
        "institution",
        "edition",
        "month",
        "day",
        "editor",
        "urldate",
        "medium",
        "year",
        "volume",
        "number",
        "pages",
        "doi",
        "url",
    }
    extra = {key: value for key, value in fields.items() if key not in standard_keys}
    return Reference(
        entry_type=entry_type,
        cite_key=cite_key,
        title=fields.get("title"),
        authors=authors,
        journal=fields.get("journal"),
        booktitle=fields.get("booktitle"),
        publisher=fields.get("publisher"),
        place=fields.get("address"),
        institution=fields.get("institution"),
        edition=fields.get("edition"),
        month=fields.get("month"),
        day=fields.get("day"),
        editors=_split_authors(fields.get("editor", "")),
        accessed_date=fields.get("urldate"),
        medium=fields.get("medium"),
        year=fields.get("year"),
        volume=fields.get("volume"),
        issue=issue,
        report_number=report_number,
        pages=fields.get("pages"),
        doi=fields.get("doi"),
        url=fields.get("url"),
        extra_fields=extra,
    )


def _split_authors(raw: str) -> list[str]:
    if not raw:
        return []
    return [author.strip() for author in raw.split(" and ") if author.strip()]


__all__ = ["BibTeXError", "parse_bibtex_entry", "reference_to_bibtex"]
